src/Markup/Avalonia.Markup.Xaml.Loader/CompilerExtensions/Transformers/AvaloniaXamlIlOptionMarkupExtensionTransformer.cs

```diff
@@ -0,0 +1,405 @@
+using XamlX.TypeSystem;
+
+namespace Avalonia.Markup.Xaml.XamlIl.CompilerExtensions.Transformers;
+
+internal class AvaloniaXamlIlOptionMarkupExtensionTransformer : IXamlAstTransformer
+{
+    private const string OnFqn = "Avalonia.Markup.Xaml:Avalonia.Markup.Xaml.MarkupExtensions.On";
+
+    public IXamlAstNode Transform(AstTransformationContext context, IXamlAstNode node)
+    {
+        if (node is XamlMarkupExtensionNode
+            {
+                Value: XamlAstObjectNode { Type: XamlAstClrTypeReference { Type: { } type } } objectNode
+            } markupExtensionNode
+            && type.FindMethods(m => m.IsPublic && m.Parameters.Count is 1 or 2 && m.ReturnType == context.Configuration.WellKnownTypes.Boolean && m.Name == "ShouldProvideOption").ToArray() is { } methods
+            && methods.Any())
+        {
+            var optionAttribute = context.GetAvaloniaTypes().MarkupExtensionOptionAttribute;
+            var defaultOptionAttribute = context.GetAvaloniaTypes().MarkupExtensionDefaultOptionAttribute;
+
+            var typeArgument = type.GenericArguments?.FirstOrDefault();
+
+            IXamlAstValueNode defaultValue = null;
+            var values = new List<OptionsMarkupExtensionBranch>();
+
+            if (objectNode.Arguments.FirstOrDefault() is { } argument)
+            {
+                var hasDefaultProp = objectNode.Type.GetClrType().GetAllProperties().Any(p =>
+                    p.CustomAttributes.Any(a => a.Type == defaultOptionAttribute));
+                if (hasDefaultProp)
+                {
+                    if (objectNode.Arguments.Count > 1)
+                    {
+                        throw new XamlParseException("Options MarkupExtensions allow only single argument", objectNode);
+                    }
+
+                    defaultValue = TransformNode(new[] { argument }, typeArgument, objectNode);
+                    objectNode.Arguments.Remove(argument);
+                }
+            }
+
+            foreach (var extProp in objectNode.Children.OfType<XamlAstXamlPropertyValueNode>().ToArray())
+            {
+                if (!extProp.Values.Any())
+                {
+                    continue;
+                }
+
+                var shouldRemoveProp = false;
+                var onObjs = extProp.Values.OfType<XamlAstObjectNode>()
+                    .Where(o => o.Type.GetClrType().GetFqn() == OnFqn).ToArray();
+                if (onObjs.Any())
+                {
+                    shouldRemoveProp = true;
+                    foreach (var onObj in onObjs)
+                    {
+                        var optionsPropNode = onObj.Children.OfType<XamlAstXamlPropertyValueNode>()
+                            .SingleOrDefault(v => v.Property.GetClrProperty().Name == "Options")
+                            ?.Values.Single();
+                        var options = (optionsPropNode as XamlAstTextNode)?.Text?.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries)
```

Comment:
Do we have anything build-in to convert string node to an array? Maybe a clr attribute?
If not, I would rather not add it in this PR.